File: bridge/test_f9_canonical_digest.py
"""Python side of the cross-language F9 canonical digest golden gate."""

from __future__ import annotations

import copy
import json
import unittest
from pathlib import Path

from contracts.evidence_agent import canonical_json_text, canonical_sha256, snapshot_material


FIXTURE = Path(__file__).parent / "fixtures" / "f9-canonical-digest-golden.json"


def materialize_integers(value: object, pointers: list[str]) -> object:
    value = copy.deepcopy(value)
    for pointer in pointers:
        tokens = pointer.lstrip("/").split("/")
        target = value
        for token in tokens[:-1]:
            target = target[int(token)] if isinstance(target, list) else target[token]
        last = tokens[-1]
        if isinstance(target, list):
            target[int(last)] = int(target[int(last)], 10)
        else:
            target[last] = int(target[last], 10)
    return value


class CanonicalDigestGoldenTest(unittest.TestCase):
    def test_python_matches_all_golden_vectors_losslessly(self) -> None:
        fixture = json.loads(FIXTURE.read_text(encoding="utf-8"))
        self.assertEqual(fixture["schema_version"], "tilesim.bridge.canonical_digest_golden.v1")
        for case in fixture["cases"]:
            with self.subTest(case=case["id"]):
                value = materialize_integers(case["value"], case["integer_json_pointers"])
                self.assertEqual(canonical_json_text(value), case["expected_canonical_json"])
                self.assertEqual(canonical_sha256(value), case["expected_sha256"])

    def test_binary_float_is_rejected(self) -> None:
        with self.assertRaisesRegex(TypeError, "floating-point"):
            canonical_json_text({"unsafe": 0.1})

    def test_snapshot_material_uses_request_identity_and_excludes_untrusted_fields(self) -> None:
        request = {
            "schema_version": "tilesim.bridge.evidence_agent_request.v1",
            "schema_set_revision": "sha256:" + "1" * 64,
            "run_id": "run-golden",
            "structured_report_schema_identity": "tilesim.web.structured-performance-report.v2",
            "snapshot_reference": {"schema_version": "tilesim.bridge.evidence_snapshot_reference.v1"},
            "artifact_allow_list": [{"bytes": 9_007_199_254_740_993_123}],
            "locale": "zh-CN",
            "task_kind": "explain_p99",
            "client_request_id": "client-one",
            "user_question": {"content": "first"},
        }
        material = snapshot_material(request)
        self.assertEqual(material["schema_version"], request["schema_version"])
        changed = copy.deepcopy(request)
        changed.update(locale="en-US", task_kind="explain_tail", client_request_id="client-two")
        changed["user_question"] = {"content": "second"}
        self.assertEqual(canonical_sha256(material), canonical_sha256(snapshot_material(changed)))


if __name__ == "__main__":
    unittest.main()

File: bridge/test_f9_canonical_digest.mjs
import assert from "node:assert/strict";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.dirname(fileURLToPath(import.meta.url));
const fixture = JSON.parse(fs.readFileSync(path.join(root, "fixtures", "f9-canonical-digest-golden.json"), "utf8"));

function compareCodePoints(left, right) {
  const a = Array.from(left, (character) => character.codePointAt(0));
  const b = Array.from(right, (character) => character.codePointAt(0));
  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

function canonical(value) {
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "bigint") return value.toString(10);
  if (typeof value === "number") {
    if (!Number.isFinite(value) || (Number.isInteger(value) && !Number.isSafeInteger(value))) {
      throw new TypeError("Canonical JSON requires a finite lossless number.");
    }
    return Object.is(value, -0) ? "0" : JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object") {
    return `{${Object.keys(value)
      .sort(compareCodePoints)
      .map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`)
      .join(",")}}`;
  }
  throw new TypeError(`Unsupported canonical type: ${typeof value}`);
}

function materializeIntegers(value, pointers) {
  value = structuredClone(value);
  for (const pointer of pointers) {
    const tokens = pointer.slice(1).split("/");
    let target = value;
    for (const token of tokens.slice(0, -1)) target = target[Array.isArray(target) ? Number(token) : token];
    const last = tokens.at(-1);
    target[Array.isArray(target) ? Number(last) : last] = BigInt(target[Array.isArray(target) ? Number(last) : last]);
  }
  return value;
}

assert.equal(fixture.schema_version, "tilesim.bridge.canonical_digest_golden.v1");
for (const vector of fixture.cases) {
  const value = materializeIntegers(vector.value, vector.integer_json_pointers);
  const text = canonical(value);
  assert.equal(text, vector.expected_canonical_json, vector.id);
  assert.equal(
    `sha256:${crypto.createHash("sha256").update(text, "utf8").digest("hex")}`,
    vector.expected_sha256,
    vector.id,
  );
}

console.log(`F9 canonical digest golden vectors passed in Node (${fixture.cases.length} cases).`);
